docs/import/add_disaster.js

```diff
@@ -0,0 +1,292 @@
+import {eeStatePrefixLength, legacyStateDir} from '../ee_paths.js';
+import {getFirestoreRoot} from '../firestore_document.js';
+
+export {enableWhenReady, SENTINEL_NEW_DISASTER_VALUE, toggleState};
+// Visible for testing
+export {
+  addDisaster,
+  createAssetPickers,
+  createOptionFrom,
+  disasters,
+  emptyCallback,
+  getAssetsFromEe,
+  stateAssets,
+  writeNewDisaster,
+};
+
+// Currently a map of disaster name to states. This pulls once on firebase
+// authentication and then makes local updates afterwards so we don't need to
+// wait on firebase writes to read new info.
+const disasters = new Map();
+
+// Map of state to list of known assets
+const stateAssets = new Map();
+
+const SENTINEL_NEW_DISASTER_VALUE = 'NEW DISASTER';
+
+// TODO: general reminder to add loading indicators for things like creating
+// new state asset folders, etc.
+
+/**
+ * Populates the disaster picker with disasters from firestore, and enables
+ * the ability to add a new disaster.
+ * @return {Promise<firebase.firestore.QuerySnapshot>}
+ */
+function enableWhenReady() {
+  // enable (currently hidden) add disaster button now that firestore is ready.
+  const addDisasterButton = $('#add-disaster-button');
+  addDisasterButton.prop('disabled', false);
+  addDisasterButton.on('click', addDisaster);
+
+  // populate disaster picker.
+  return getFirestoreRoot()
+      .collection('disaster-metadata')
+      .get()
+      .then((querySnapshot) => {
+        const disasterPicker = $('#disaster');
+        querySnapshot.forEach((doc) => {
+          const name = doc.id;
+          disasterPicker.prepend(createOptionFrom(name));
+          disasters.set(name, doc.data().states);
+        });
+
+        disasterPicker.on('change', () => toggleState(disasterPicker.val()));
+        const mostRecent = querySnapshot.docs[querySnapshot.size - 1].id;
+        disasterPicker.val(mostRecent).trigger('change');
+      });
+}
+
+/**
+ * Switch between the page creating a new disaster and the page editing a
+ * current disaster.
+ * @param {String} disaster
+ * @return {Promise<void>} completes when we've finished filling all state
+ * pickers.
+ */
+function toggleState(disaster) {
+  if (disaster === SENTINEL_NEW_DISASTER_VALUE) {
+    $('#new-disaster').show();
+    $('#selected-disaster').hide();
+  } else {
+    const states = disasters.get(disaster);
+    const statesToFetch = [];
+    for (const state of states) {
+      if (!stateAssets.has(state)) statesToFetch.push(state);
+    }
+    // TODO: add functionality to re-pull all cached states from ee without
+    // reloading the page.
+    let assetPickersDone = Promise.resolve();
```

Comment:
Switching to a disaster should also pull from Firestore, right? Can you leave a TODO or skeleton of how/where that would happen? I think it will complicate this logic enough that it might be worth taking that into account here.

Reply:
My basic idea is we keep the entire disaster-metadata document for each disaster in the disasters map values the way we're keeping the states right now. That way, we never have to read from firestore except for the first time in the beginning so we don't have to wait and make sure writes are done. I don't think it'll be too too much data and never having to do reads sounds pretty nice. In that case, I don't think there's too much to do here right now. Thoughts?